File: app/src/main/java/com/asianliftbd/staff/CallbackProject.java
package com.asianliftbd.staff;

import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.RelativeLayout;
import android.widget.ScrollView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.asianliftbd.staff.db.HReport;
import com.asianliftbd.staff.user.UCallback;
import com.asianliftbd.staff.user.UTC;
import com.firebase.ui.database.FirebaseRecyclerAdapter;
import com.firebase.ui.database.FirebaseRecyclerOptions;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

public class CallbackProject extends AppCompatActivity {
    private String name;
    private TextView Empty,Project;
    private ScrollView l;
    private RecyclerView rv;
    private DatabaseReference ref,tref;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        overridePendingTransition(android.R.anim.fade_in, android.R.anim.fade_out);
        setContentView(R.layout.callback_project);
        Bundle extras = getIntent().getExtras();
        if (extras != null) {
            name=extras.getString("key");
        }
        ref = FirebaseDatabase.getInstance().getReference().child("callback/details/"+name);
        tref = FirebaseDatabase.getInstance().getReference().child("callback/total/"+name);
        ref.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot s) {
                int count= (int) s.getChildrenCount();
                UTC u=new UTC(name,count);
                tref.setValue(u);
            }
            @Override
            public void onCancelled(@NonNull DatabaseError e) {}
        });
        l=findViewById(R.id.l);
        Empty=findViewById(R.id.empty);
        Project=findViewById(R.id.project);
        Project.setText(name);
        rv=findViewById(R.id.rv);
        LinearLayoutManager LM = new LinearLayoutManager(CallbackProject.this);
        LM.setReverseLayout(true);
        LM.setStackFromEnd(true);
        rv.setLayoutManager(LM);
    }
    @Override
    public void onStart() {
        super.onStart();
        ref.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot d) {
                if (!d.exists()) {
                    Empty.setVisibility(View.VISIBLE);
                    l.setVisibility(View.GONE);
                } else {
                    l.setVisibility(View.VISIBLE);
                    Empty.setVisibility(View.GONE);
                    FirebaseRecyclerOptions<UCallback> options =
                            new FirebaseRecyclerOptions.Builder<UCallback>()
                                    .setQuery(ref, UCallback.class)
                                    .build();
                    FirebaseRecyclerAdapter<UCallback, HReport> a = new FirebaseRecyclerAdapter<UCallback, HReport>(options) {
                        @Override
                        protected void onBindViewHolder(@NonNull HReport vh, int position, @NonNull UCallback ut) {
                            vh.setDetails(ut.getName(),ut.getDetails(),ut.getDate());
                        }
                        @NonNull
                        @Override
                        public HReport onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
                            View view = LayoutInflater.from(parent.getContext())
                                    .inflate(R.layout.m_report, parent, false);
                            return new HReport(view);
                        }
                    };
                    rv.setAdapter(a);
                    a.startListening();
                }
            }
            @Override
            public void onCancelled(@NonNull DatabaseError e) {
            }
        });
    }
    public void Back(View v){finish();}
    public void Report(View v){startActivity(new Intent(this,Report.class));}
}

File: app/src/main/java/com/asianliftbd/staff/activity/CallbackProject.kt
package com.asianliftbd.staff.activity

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.asianliftbd.staff.R
import com.asianliftbd.staff.databinding.ActivityCallbackProjectBinding
import com.asianliftbd.staff.model.CallbackModel
import com.asianliftbd.staff.viewHolder.CallbackHolder
import com.firebase.ui.database.FirebaseRecyclerAdapter
import com.firebase.ui.database.FirebaseRecyclerOptions
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener

class CallbackProject : AppCompatActivity() {
    private lateinit var binding: ActivityCallbackProjectBinding

    private lateinit var recyclerView: RecyclerView
    private val linearLayoutManager = LinearLayoutManager(this)
    private lateinit var emptyTextView: TextView

    private lateinit var project: String
    private lateinit var ref: DatabaseReference

    private val uid = FirebaseAuth.getInstance().currentUser!!.uid

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityCallbackProjectBinding.inflate(layoutInflater)
        setContentView(binding.root)

        setSupportActionBar(binding.toolbar)

        val extras = intent.extras
        if (extras != null) {
            project = extras.getString("project").toString()
            supportActionBar?.title = project
            ref = FirebaseDatabase.getInstance().getReference("callback/$project")
        } else {
            finish()
            Toast.makeText(this, "Error Occurred! Try again", Toast.LENGTH_SHORT).show()
        }

        recyclerView = binding.recyclerView
        emptyTextView = binding.emptyTextView

        recyclerView.run {
            layoutManager = linearLayoutManager
        }
    }

    public override fun onStart() {
        super.onStart()
        ref.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(d: DataSnapshot) {
                if (!d.exists()) {
                    recyclerView.visibility = View.GONE
                    emptyTextView.visibility = View.VISIBLE
                } else {
                    emptyTextView.visibility = View.GONE
                    recyclerView.visibility = View.VISIBLE
                    val options = FirebaseRecyclerOptions.Builder<CallbackModel>()
                        .setQuery(ref.orderByChild("dbdate"), CallbackModel::class.java)
                        .build()
                    val recyclerAdapter: FirebaseRecyclerAdapter<CallbackModel, CallbackHolder> =
                        object : FirebaseRecyclerAdapter<CallbackModel, CallbackHolder>(options) {
                            override fun onBindViewHolder(
                                vh: CallbackHolder,
                                position: Int,
                                ut: CallbackModel
                            ) {
                                vh.setDetails(applicationContext, ut.name, ut.details, ut.date, project, getRef(position).key.toString(), uid, ut.uid)
                            }

                            override fun onCreateViewHolder(
                                parent: ViewGroup,
                                viewType: Int
                            ): CallbackHolder {
                                val view = LayoutInflater.from(parent.context)
                                    .inflate(R.layout.model_transaction, parent, false)
                                return CallbackHolder(view)
                            }
                        }
                    recyclerView.adapter = recyclerAdapter
                    recyclerAdapter.startListening()
                }
            }

            override fun onCancelled(e: DatabaseError) {}
        })
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.menu_add, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.nav_add -> {
                startActivity(Intent(applicationContext, CallbackNew::class.java))
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onSupportNavigateUp(): Boolean {
        onBackPressedDispatcher.onBackPressed()
        return true
    }
}